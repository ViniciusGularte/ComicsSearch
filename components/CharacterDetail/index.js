import React, { useState } from "react";
import {
  Container,
  CharacterInfoContent,
  CharacterInfoText,
  CharacterImg,
  CharacterDescription,
  ContainerButton,
  ContainerTextAndButtons,
  CharacterInfoInput,
  CharacterInfoInputSelect,
  ButtonEdit,
  ButtonFavorite,
  ContainerArrowGoBack,
} from "./styles";
import { faStar, faEdit, faArrowLeft } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { useSelector, useDispatch } from "react-redux";
import { addCharacter } from "../../store/characters/action";
import customUseEffectUpdate from "../../hooks/customUseEffect";
import { useRouter } from "next/router";
import NotificationManager from "../../widget/notifications";
const CharacterCard = ({ characterInfo }) => {
  const dispatch = useDispatch();
  const router = useRouter();

  const stateCharacters = useSelector((state) => state.characters);
  const [edit, setEdit] = useState(false);
  const [form, setForm] = useState({
    id: characterInfo.results.id,
    name:
      stateCharacters[characterInfo.results.id] &&
      stateCharacters[characterInfo.results.id].name
        ? stateCharacters[characterInfo.results.id].name
        : characterInfo.results.name,
    gender:
      stateCharacters[characterInfo.results.id] &&
      stateCharacters[characterInfo.results.id].gender
        ? stateCharacters[characterInfo.results.id].gender
        : characterInfo.results.gender,
    real_name:
      stateCharacters[characterInfo.results.id] &&
      stateCharacters[characterInfo.results.id].real_name
        ? stateCharacters[characterInfo.results.id].real_name
        : characterInfo.results.real_name,
    aliases:
      stateCharacters[characterInfo.results.id] &&
      stateCharacters[characterInfo.results.id].aliases
        ? stateCharacters[characterInfo.results.id].aliases
        : characterInfo.results.aliases,
    birth:
      stateCharacters[characterInfo.results.id] &&
      stateCharacters[characterInfo.results.id].birth
        ? stateCharacters[characterInfo.results.id].birth
        : characterInfo.results.birth,
    image: characterInfo.results.image.medium_url,
  });
  customUseEffectUpdate(() => {
    const timeout = setTimeout(() => {
      dispatch(
        addCharacter(
          form.id,
          form.name,
          form.gender,
          form.real_name,
          form.aliases,
          form.birth,
          form.image,
          stateCharacters[form.id] && stateCharacters[form.id].is_favorite
            ? true
            : false
        )
      );
    }, 4000);
    return () => clearTimeout(timeout);
  }, [form]);
  const toogleEdit = () => {
    setEdit(!edit);
  };
  const changeFavorite = () => {
    if (
      stateCharacters[form.id] &&
      stateCharacters[form.id].is_favorite === true
    ) {
      NotificationManager("warning", "ðŸ˜”	 Removed from favorites");
      dispatch(
        addCharacter(
          form.id,
          form.name,
          form.gender,
          form.real_name,
          form.aliases,
          form.birth,
          form.image,
          false
        )
      );
    } else {
      NotificationManager("success", " ðŸ˜Š	 Added to favorites");

      dispatch(
        addCharacter(
          form.id,
          form.name,
          form.gender,
          form.real_name,
          form.aliases,
          form.birth,
          form.image,
          true
        )
      );
    }
  };
  const sendHome = (e) => {
    router.push("/");
  };

  const changeInput = (e) => {
    setForm({ ...form, [e.target.name]: e.target.value });
  };
  return (
    <Container>
      <ContainerArrowGoBack>
        <FontAwesomeIcon
          onClick={sendHome}
          color={"blue"}
          size={"1x"}
          icon={faArrowLeft}
        />
      </ContainerArrowGoBack>

      <CharacterImg src={characterInfo.results.image.medium_url} />
      <ContainerTextAndButtons>
        {edit ? (
          <CharacterInfoContent>
            <CharacterInfoInput
              onChange={changeInput}
              name="name"
              placeholder="Name"
              value={form.name}
            />
            <CharacterInfoInput
              onChange={changeInput}
              placeholder="Real name"
              name="real_name"
              value={form.real_name}
            />
            <CharacterInfoInput
              onChange={changeInput}
              placeholder="Birth"
              name="birth"
              type="date"
              value={form.birth}
            />
            <CharacterInfoInputSelect
              onChange={changeInput}
              placeholder="Gender"
              name="gender"
              type="select"
              value={form.gender}
            >
              <option value="0">Woman</option>
              <option value="1">Man</option>
            </CharacterInfoInputSelect>
            <CharacterInfoInput
              onChange={changeInput}
              placeholder="Aliases"
              name="aliases"
              value={form.aliases}
            />
          </CharacterInfoContent>
        ) : (
          <CharacterInfoContent>
            <CharacterInfoText>Name: {form.name}</CharacterInfoText>
            <CharacterInfoText>Real Name: {form.real_name}</CharacterInfoText>
            <CharacterInfoText>
              Birth: {form.birth ? form.birth : "--"}
            </CharacterInfoText>
            <CharacterInfoText>
              Gender: {form.gender === 1 ? "Man" : "Woman"}
            </CharacterInfoText>
            <CharacterInfoText>Aliases: {form.aliases}</CharacterInfoText>
          </CharacterInfoContent>
        )}
        <ContainerButton>
          <ButtonEdit onClick={toogleEdit}>
            <FontAwesomeIcon color={"white"} size={"1x"} icon={faEdit} />
          </ButtonEdit>
          <ButtonFavorite
            is_favorite={
              stateCharacters[form.id] && stateCharacters[form.id].is_favorite
            }
            onClick={changeFavorite}
          >
            <FontAwesomeIcon color={"white"} size={"1x"} icon={faStar} />
          </ButtonFavorite>
        </ContainerButton>
      </ContainerTextAndButtons>
      <CharacterDescription
        dangerouslySetInnerHTML={{
          __html: characterInfo.results.description,
        }}
      />
    </Container>
  );
};
export default CharacterCard;
