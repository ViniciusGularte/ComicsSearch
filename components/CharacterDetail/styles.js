import styled from "styled-components";
const Container = styled.div`
  background-color: ${(props) => props.theme.surface};
  display: flex;
  flex: 1;
  align-items: center;
  flex-direction: column;
  box-shadow: 0px 0px 22px 0px rgba(148, 148, 148, 0.51);
  padding: 32px;
`;
const CharacterInfoContent = styled.div`
  display: flex;
  align-self: flex-start;
  flex: 1;
  flex-direction: column;
`;
const CharacterInfoText = styled.div`
  display: flex;
  font-size: 18px;
  padding: 10px 10px 10px 5px;

  flex: 1;
  margin-bottom: 4px;
`;
const CharacterInfoInput = styled.input`
  display: flex;
  font-size: 18px;
  padding: 10px 10px 10px 5px;
  display: block;
  width: 300px;
  border: none;
  flex: 1;
  margin-bottom: 4px;
`;
const CharacterInfoInputSelect = styled.select`
  display: flex;
  background-color: white;
  font-size: 18px;
  padding: 10px 10px 10px 5px;
  display: block;
  width: 300px;
  border: none;
  flex: 1;
  margin-bottom: 4px;
`;
const CharacterDescription = styled.div`
  display: flex;
  align-self: flex-start;
  flex-direction: column;
  flex: 1;
  max-width: 60vw;
  a {
    color: inherit;
    text-decoration: inherit;
    pointer-events: none;
    cursor: vertical-text;
  }
  figure {
    margin: 0;
    a {
      img {
        @media screen and (min-width: 20em) {
          max-width: 12.5em;
        }
        @media screen and (min-width: 40em) {
          max-width: 18.5em;
        }
        @media screen and (min-width: 60em) {
          max-width: 20.5em;
        }
      }
    }
  }
  @media screen and (min-width: 60em) {
    max-width: calc(80% - 1rem);
  }
  @media screen and (min-width: 40em) {
    max-width: calc(100% - 1rem);
  }
`;
const CharacterAge = styled.div`
  display: flex;
  flex: 1;
`;
const CharacterImg = styled.img`
  object-fit: contain;
  display: flex;
  flex-direction: row;
  border-radius: 10%;
  height: 150px;
  border: 3px solid #eee;
  box-shadow: 0px 0px 22px 0px rgba(148, 148, 148, 0.51);
  margin-bottom: 16px;
`;
const ContainerTextAndButtons = styled.div`
  display: flex;
  flex: 1;
  flex-direction: row;
  @media screen and (max-width: 40em) {
    flex-direction: column;
  }
`;
const ContainerButton = styled.div`
  display: flex;
  align-self: baseline;
  flex-direction: column;
  justify-content: space-between;
  @media screen and (max-width: 40em) {
    flex-direction: row;
  }
`;
const Button = styled.button`
  flex: 1;
  padding: 10px;
  border-radius: 50%;
  margin: 4px;
  border: none;
  cursor: pointer;
`;
const ButtonEdit = styled(Button)`
  background-color: ${(props) => props.theme.warning};
`;
const ButtonFavorite = styled(Button)`
  background-color: ${(props) =>
    props.is_favorite ? props.theme.warning : null};
`;
const ContainerArrowGoBack = styled.div`
  flex: 1;
  display: flex;
  flex-direction: row;
  align-self: flex-start;
  svg {
    cursor: pointer;
    transition: transform 0.2s;
    &:hover {
      transform: scale(1.4);
    }
  }
`;
export {
  Container,
  CharacterInfoText,
  CharacterInfoInputSelect,
  ContainerTextAndButtons,
  CharacterInfoInput,
  CharacterAge,
  CharacterImg,
  CharacterDescription,
  CharacterInfoContent,
  ContainerButton,
  ButtonEdit,
  ButtonFavorite,
  ContainerArrowGoBack,
};
