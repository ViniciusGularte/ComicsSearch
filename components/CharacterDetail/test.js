import React from "react";
import Enzyme, { shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import CharacterDetail from "./index";
const initialState = {
  characters: {},
};
const characterInfo = {
  results: {
    id: "545",
    name: "vini",
    image: {
      medium_url: "",
    },
  },
};
Enzyme.configure({ adapter: new Adapter() });

describe("CharacterDetail", () => {
  const mockStore = configureStore();
  let store;
  it("renders correctly", () => {
    store = mockStore(initialState);
    const wrapper = shallow(
      <Provider store={store}>
        <CharacterDetail characterInfo={characterInfo} />{" "}
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });
  it("renders character name", () => {
    store = mockStore(initialState);
    const wrapper = shallow(
      <Provider store={store}>
        <CharacterDetail characterInfo={characterInfo} />{" "}
      </Provider>
    );
  });
});
