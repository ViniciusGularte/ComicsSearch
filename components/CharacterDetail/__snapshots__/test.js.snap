// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterDetail renders correctly 1`] = `ShallowWrapper {}`;
